woocommerce/woocommerce-gutenberg-products-block
# Refactor checkout context provider to be more flexible and implement CartProvider

Contributes towards: #1882

The plan I had in mind when creating the `CheckoutProvider` was that the logic from the different providers could be re-used for Cart as well. The idea being that we could have a`CartProvider` that wraps the cart and exposing similar interface for the cart as that exposed to the checkout. This would avoid duplication in code since Cart and Checkout share so many components and logic.

The need for this became more apparent with @senadir 's work on #1882 (see #1978 and the discussion there).

So in this pull the following is implemented:

- The interface in `CheckoutProvider` was extracted and moved to `CheckoutStateProvider` (not settled on the name but I've left it for now, suggestions welcome).
- `CheckoutProvider` now implements `CheckoutStateProvider` and the various providers wrapping the checkout block components.
- `CartProvider` has been created that implements `CheckoutStateProvider` and _just_ the `ShippingDataProvider` since that's all Cart needs currently.
- `useCheckoutContext` exposes the interface from `CheckoutStateProvider` so it can be used in any component that is implemented in either the cart or block. However, the name will likely be confusing so ideas are welcome for renaming this to something more generic for use in components implemented by checkout and/or cart.
- I cleaned up some type-defs (not exhaustive, just a few).
- We have `EditorProvider` for exposing the `isEditor` boolean to components, so I dropped `isEditor` flag from checkout context and refactored all implementations of `isEditor` coming from `useCheckoutContext` to be pulled from `useEditorContext` instead.

With this pull merged, we can continue with switching hook implementations in cart and checkout components to use the relevant context hooks instead.  @senadir should also be able to finish off #1882.

## To test

This impacts both cart and checkout blocks (mostly checkout). So testing should involve:

- verify frontend and editor instance of both blocks work as expected (with no console errors).
- verify inserting either block in a new block-editor instance inserts the block as expected and there are no errors in the console.
